// postcss-preset-env/index.es.js
import postcssApply from 'postcss-apply';
import postcssInitial from 'postcss-initial';
import postcssColorHexAlpha from 'postcss-color-hex-alpha';
import postcssColorModFunction from 'postcss-color-mod-function';
import postcssColorRebeccapurple from 'postcss-color-rebeccapurple';
import postcssCustomSelectors from 'postcss-custom-selectors';
import postcssFocusVisible from 'postcss-focus-visible';
import postcssFontVariant from 'postcss-font-variant';
import postcssFontFamilySystemUi from 'postcss-font-family-system-ui';
import postcssImageSetPolyfill from '@csstools/postcss-image-set-function';
import postcssLogical from 'postcss-logical';
import postcssNesting from 'postcss-nesting';
import postcssReplaceOverflowWrap from 'postcss-replace-overflow-wrap';
import postcssCustomProperties from 'postcss-custom-properties';
import postcssCustomMedia from 'postcss-custom-media';
import postcssMediaMinmax from 'postcss-media-minmax';
import postcssPageBreak from 'postcss-page-break';
import postcssPseudoClassAnyLink from 'postcss-pseudo-class-any-link';
import postcssAttributeCaseInsensitive from 'postcss-attribute-case-insensitive';
import postcssSelectorMatches from 'postcss-selector-matches';
import postcssSelectorNot from 'postcss-selector-not';
import postcssDirPseudoClass from 'postcss-dir-pseudo-class';
import { features, feature } from 'caniuse-lite';
import browserslist from 'browserslist';
import cssdb from 'cssdb';
import postcss from 'postcss';

// tooling

// postcss plugins ordered by specification id
var plugins = {
	'break-properties': postcssPageBreak,
	'css-apply-rule': postcssApply,
	'css-cascade-all-shorthand': postcssInitial,
	'css-color-hex-notation': postcssColorHexAlpha,
	'css-color-modifying-colors': postcssColorModFunction,
	'css-color-valdef-color-rebeccapurple': postcssColorRebeccapurple,
	'css-extensions-custom-selectors': postcssCustomSelectors,
	'css-fonts-propdef-font-variant': postcssFontVariant,
	'css-fonts-system-ui-def': postcssFontFamilySystemUi,
	'css-images-image-set-notation': postcssImageSetPolyfill,
	'css-logical': postcssLogical,
	'css-nesting': postcssNesting,
	'css-text-overflow-wrap-property': postcssReplaceOverflowWrap,
	'css-variables': postcssCustomProperties,
	'mediaqueries-custom-mq': postcssCustomMedia,
	'mediaqueries-mq-ranges': postcssMediaMinmax,
	'selectors-any-link-pseudo': postcssPseudoClassAnyLink,
	'selectors-attribute-case': postcssAttributeCaseInsensitive,
	'selectors-dir-pseudo': postcssDirPseudoClass,
	'selectors-focus-visible-pseudo': postcssFocusVisible,
	'selectors-matches': postcssSelectorMatches,
	'selectors-negation': postcssSelectorNot
};

// return a list of features to be inserted before or after cssdb features
function getTransformedInsertions(insertions, placement) {
	return Object.keys(insertions).map(function (specificationId) {
		return [].concat(insertions[specificationId]).map(function (plugin) {
			return {
				[placement]: true,
				plugin,
				specificationId
			};
		});
	}).reduce(function (array, feature$$1) {
		return array.concat(feature$$1);
	}, []);
}

// tooling

// return a list of browsers that do not support the feature
function getUnsupportedBrowsersByFeature(feature$$1) {
	var caniuseFeature = features[feature$$1];

	// if feature support can be determined
	if (caniuseFeature) {
		var stats = feature(caniuseFeature).stats;

		// return an array of browsers and versions that do not support the feature
		var results = Object.keys(stats).reduce(function (browsers, browser) {
			return browsers.concat(Object.keys(stats[browser]).filter(function (version) {
				return stats[browser][version].indexOf('y') !== 0;
			}).map(function (version) {
				return `${browser} ${version}`;
			}));
		}, []);

		return results;
	} else {
		// otherwise, return that the feature does not work in any browser
		return ['> 0%'];
	}
}

// specification ids ordered by execution
var specificationIdsByExecutionOrder = ['css-variables', 'postcss-apply', 'css-images-image-set-notation', 'css-logical', 'css-nesting', 'mediaqueries-5-custom-mq', 'mediaqueries-mq-ranges', 'css-extensions-custom-selectors', 'selectors-attribute-case', 'css-color-valdef-color-rebeccapurple', 'css-color-hex-notation', 'css-color-modifying-colors', 'css-fonts-system-ui-def', 'css-fonts-propdef-font-variant', 'css-cascade-all-shorthand', 'selectors-matches', 'selectors-negation', 'selectors-any-link-pseudo', 'selectors-dir-pseudo', 'css-text-overflow-wrap-property'];

// tooling

// plugin
var index = postcss.plugin('postcss-preset-env', function (opts) {
	// initialize options
	var features$$1 = Object(Object(opts).features);
	var insertBefore = Object(Object(opts).insertBefore);
	var insertAfter = Object(Object(opts).insertAfter);
	var browsers = Object(opts).browsers;
	var stage = 'stage' in Object(opts) ? opts.stage === false ? 6 : parseInt(opts.stage) || 0 : 3;

	// polyfillable features (those with an available postcss plugin)
	var polyfillableFeatures = cssdb.concat(
	// additional features to be inserted before cssdb features
	getTransformedInsertions(insertBefore, 'insertBefore'),
	// additional features to be inserted after cssdb features
	getTransformedInsertions(insertAfter, 'insertAfter')).filter(
	// inserted features or features with an available postcss plugin
	function (feature$$1) {
		return feature$$1.insertBefore || feature$$1.specificationId in plugins;
	}).sort(
	// features sorted by execution order and then insertion order
	function (a, b) {
		return specificationIdsByExecutionOrder.indexOf(a.specificationId) - specificationIdsByExecutionOrder.indexOf(b.specificationId) || (a.insertBefore ? -1 : b.insertBefore ? 1 : 0) || (a.insertAfter ? 1 : b.insertAfter ? -1 : 0);
	}).map(
	// polyfillable features as an object
	function (feature$$1) {
		// target browsers for the polyfill
		var unsupportedBrowsers = getUnsupportedBrowsersByFeature(feature$$1.caniuse);

		return feature$$1.insertBefore || feature$$1.insertAfter ? {
			browsers: unsupportedBrowsers,
			plugin: feature$$1.plugin,
			specificationId: `${feature$$1.insertBefore ? 'before' : 'after'}-${feature$$1.specificationId}`,
			stage: 6
		} : {
			browsers: unsupportedBrowsers,
			plugin: plugins[feature$$1.specificationId],
			specificationId: feature$$1.specificationId,
			stage: feature$$1.stage
		};
	});

	// staged features (those at or above the selected stage)
	var stagedFeatures = polyfillableFeatures.filter(function (feature$$1) {
		return feature$$1.specificationId in features$$1 ? features$$1[feature$$1.specificationId] : feature$$1.stage >= stage;
	}).map(function (feature$$1) {
		return {
			browsers: feature$$1.browsers,
			plugin: typeof feature$$1.plugin.process === 'function' ? features$$1[feature$$1.specificationId] === true ? feature$$1.plugin() : feature$$1.plugin(features$$1[feature$$1.specificationId]) : feature$$1.plugin,
			specificationId: feature$$1.specificationId
		};
	});

	return function (root, result) {
		// browsers supported by the configuration
		var supportedBrowsers = browserslist(browsers, {
			path: result.root.source && result.root.source.input && result.root.source.input.file,
			ignoreUnknownVersions: true
		});

		// features supported by the stage and browsers
		var supportedFeatures = stagedFeatures.filter(function (feature$$1) {
			return supportedBrowsers.some(function (supportedBrowser) {
				return browserslist(feature$$1.browsers, {
					ignoreUnknownVersions: true
				}).some(function (polyfillBrowser) {
					return polyfillBrowser === supportedBrowser;
				});
			});
		});

		// polyfills run in execution order
		var polyfills = supportedFeatures.reduce(function (promise, feature$$1) {
			return promise.then(function () {
				return feature$$1.plugin(result.root, result);
			});
		}, Promise.resolve());

		return polyfills;
	};
});

export default index;
