// postcss-preset-env/index.cjs.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var postcssApply = _interopDefault(require('postcss-apply'));
var postcssInitial = _interopDefault(require('postcss-initial'));
var postcssColorHexAlpha = _interopDefault(require('postcss-color-hex-alpha'));
var postcssColorModFunction = _interopDefault(require('postcss-color-mod-function'));
var postcssColorRebeccapurple = _interopDefault(require('postcss-color-rebeccapurple'));
var postcssCustomSelectors = _interopDefault(require('postcss-custom-selectors'));
var postcssFocusVisible = _interopDefault(require('postcss-focus-visible'));
var postcssFontVariant = _interopDefault(require('postcss-font-variant'));
var postcssFontFamilySystemUi = _interopDefault(require('postcss-font-family-system-ui'));
var postcssImageSetPolyfill = _interopDefault(require('@csstools/postcss-image-set-function'));
var postcssLogical = _interopDefault(require('postcss-logical'));
var postcssNesting = _interopDefault(require('postcss-nesting'));
var postcssReplaceOverflowWrap = _interopDefault(require('postcss-replace-overflow-wrap'));
var postcssCustomProperties = _interopDefault(require('postcss-custom-properties'));
var postcssCustomMedia = _interopDefault(require('postcss-custom-media'));
var postcssMediaMinmax = _interopDefault(require('postcss-media-minmax'));
var postcssPageBreak = _interopDefault(require('postcss-page-break'));
var postcssPseudoClassAnyLink = _interopDefault(require('postcss-pseudo-class-any-link'));
var postcssAttributeCaseInsensitive = _interopDefault(require('postcss-attribute-case-insensitive'));
var postcssSelectorMatches = _interopDefault(require('postcss-selector-matches'));
var postcssSelectorNot = _interopDefault(require('postcss-selector-not'));
var postcssDirPseudoClass = _interopDefault(require('postcss-dir-pseudo-class'));
var caniuse = require('caniuse-lite');
var browserslist = _interopDefault(require('browserslist'));
var cssdb = _interopDefault(require('cssdb'));
var postcss = _interopDefault(require('postcss'));

// tooling

// postcss plugins ordered by specification id
var plugins = {
	'break-properties': postcssPageBreak,
	'css-apply-rule': postcssApply,
	'css-cascade-all-shorthand': postcssInitial,
	'css-color-hex-notation': postcssColorHexAlpha,
	'css-color-modifying-colors': postcssColorModFunction,
	'css-color-valdef-color-rebeccapurple': postcssColorRebeccapurple,
	'css-extensions-custom-selectors': postcssCustomSelectors,
	'css-fonts-propdef-font-variant': postcssFontVariant,
	'css-fonts-system-ui-def': postcssFontFamilySystemUi,
	'css-images-image-set-notation': postcssImageSetPolyfill,
	'css-logical': postcssLogical,
	'css-nesting': postcssNesting,
	'css-text-overflow-wrap-property': postcssReplaceOverflowWrap,
	'css-variables': postcssCustomProperties,
	'mediaqueries-custom-mq': postcssCustomMedia,
	'mediaqueries-mq-ranges': postcssMediaMinmax,
	'selectors-any-link-pseudo': postcssPseudoClassAnyLink,
	'selectors-attribute-case': postcssAttributeCaseInsensitive,
	'selectors-dir-pseudo': postcssDirPseudoClass,
	'selectors-focus-visible-pseudo': postcssFocusVisible,
	'selectors-matches': postcssSelectorMatches,
	'selectors-negation': postcssSelectorNot
};

// return a list of features to be inserted before or after cssdb features
function getTransformedInsertions(insertions, placement) {
	return Object.keys(insertions).map(function (specificationId) {
		return [].concat(insertions[specificationId]).map(function (plugin) {
			return {
				[placement]: true,
				plugin,
				specificationId
			};
		});
	}).reduce(function (array, feature) {
		return array.concat(feature);
	}, []);
}

// tooling

// return a list of browsers that do not support the feature
function getUnsupportedBrowsersByFeature(feature) {
	var caniuseFeature = caniuse.features[feature];

	// if feature support can be determined
	if (caniuseFeature) {
		var stats = caniuse.feature(caniuseFeature).stats;

		// return an array of browsers and versions that do not support the feature
		var results = Object.keys(stats).reduce(function (browsers, browser) {
			return browsers.concat(Object.keys(stats[browser]).filter(function (version) {
				return stats[browser][version].indexOf('y') !== 0;
			}).map(function (version) {
				return `${browser} ${version}`;
			}));
		}, []);

		return results;
	} else {
		// otherwise, return that the feature does not work in any browser
		return ['> 0%'];
	}
}

// specification ids ordered by execution
var specificationIdsByExecutionOrder = ['css-variables', 'postcss-apply', 'css-images-image-set-notation', 'css-logical', 'css-nesting', 'mediaqueries-5-custom-mq', 'mediaqueries-mq-ranges', 'css-extensions-custom-selectors', 'selectors-attribute-case', 'css-color-valdef-color-rebeccapurple', 'css-color-hex-notation', 'css-color-modifying-colors', 'css-fonts-system-ui-def', 'css-fonts-propdef-font-variant', 'css-cascade-all-shorthand', 'selectors-matches', 'selectors-negation', 'selectors-any-link-pseudo', 'selectors-dir-pseudo', 'css-text-overflow-wrap-property'];

// tooling

// plugin
var index = postcss.plugin('postcss-preset-env', function (opts) {
	// initialize options
	var features = Object(Object(opts).features);
	var insertBefore = Object(Object(opts).insertBefore);
	var insertAfter = Object(Object(opts).insertAfter);
	var browsers = Object(opts).browsers;
	var stage = 'stage' in Object(opts) ? opts.stage === false ? 6 : parseInt(opts.stage) || 0 : 3;

	// polyfillable features (those with an available postcss plugin)
	var polyfillableFeatures = cssdb.concat(
	// additional features to be inserted before cssdb features
	getTransformedInsertions(insertBefore, 'insertBefore'),
	// additional features to be inserted after cssdb features
	getTransformedInsertions(insertAfter, 'insertAfter')).filter(
	// inserted features or features with an available postcss plugin
	function (feature) {
		return feature.insertBefore || feature.specificationId in plugins;
	}).sort(
	// features sorted by execution order and then insertion order
	function (a, b) {
		return specificationIdsByExecutionOrder.indexOf(a.specificationId) - specificationIdsByExecutionOrder.indexOf(b.specificationId) || (a.insertBefore ? -1 : b.insertBefore ? 1 : 0) || (a.insertAfter ? 1 : b.insertAfter ? -1 : 0);
	}).map(
	// polyfillable features as an object
	function (feature) {
		// target browsers for the polyfill
		var unsupportedBrowsers = getUnsupportedBrowsersByFeature(feature.caniuse);

		return feature.insertBefore || feature.insertAfter ? {
			browsers: unsupportedBrowsers,
			plugin: feature.plugin,
			specificationId: `${feature.insertBefore ? 'before' : 'after'}-${feature.specificationId}`,
			stage: 6
		} : {
			browsers: unsupportedBrowsers,
			plugin: plugins[feature.specificationId],
			specificationId: feature.specificationId,
			stage: feature.stage
		};
	});

	// staged features (those at or above the selected stage)
	var stagedFeatures = polyfillableFeatures.filter(function (feature) {
		return feature.specificationId in features ? features[feature.specificationId] : feature.stage >= stage;
	}).map(function (feature) {
		return {
			browsers: feature.browsers,
			plugin: typeof feature.plugin.process === 'function' ? features[feature.specificationId] === true ? feature.plugin() : feature.plugin(features[feature.specificationId]) : feature.plugin,
			specificationId: feature.specificationId
		};
	});

	return function (root, result) {
		// browsers supported by the configuration
		var supportedBrowsers = browserslist(browsers, {
			path: result.root.source && result.root.source.input && result.root.source.input.file,
			ignoreUnknownVersions: true
		});

		// features supported by the stage and browsers
		var supportedFeatures = stagedFeatures.filter(function (feature) {
			return supportedBrowsers.some(function (supportedBrowser) {
				return browserslist(feature.browsers, {
					ignoreUnknownVersions: true
				}).some(function (polyfillBrowser) {
					return polyfillBrowser === supportedBrowser;
				});
			});
		});

		// polyfills run in execution order
		var polyfills = supportedFeatures.reduce(function (promise, feature) {
			return promise.then(function () {
				return feature.plugin(result.root, result);
			});
		}, Promise.resolve());

		return polyfills;
	};
});

module.exports = index;
